src/Runner/Filter/XmlTestsIterator.php

```diff
@@ -0,0 +1,142 @@
+<?php declare(strict_types=1);
+/*
+ * This file is part of PHPUnit.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace PHPUnit\Runner\Filter;
+
+use Exception;
+use LibXMLError;
+use PHPUnit\Framework\TestCase;
+use PHPUnit\Framework\TestSuite;
+use PHPUnit\Runner\PhptTestCase;
+use RecursiveFilterIterator;
+use RecursiveIterator;
+use SimpleXMLElement;
+
+/**
+ * @internal This class is not covered by the backward compatibility promise for PHPUnit
+ */
+final class XmlTestsIterator extends RecursiveFilterIterator
+{
+    /**
+     * The filter is used as a fast look for
+     * - the class name
+     * - the method name plus its optional data set description.
+     *
+     * Example: `filter[class name][method name + data set] = true;`
+     *
+     * The accept() method then can use a fast isset() to check if a test should
+     * be included or not.
+     *
+     * This works equally for phpt tests, except we hardcode the class name.
+     *
+     * @var array<string,array<string,true>>
+     */
+    private $filter = [];
+
+    /**
+     * @throws Exception
+     */
+    public function __construct(RecursiveIterator $iterator, string $xmlFile)
+    {
+        parent::__construct($iterator);
+
+        $this->setFilter($xmlFile);
+    }
+
+    public function accept(): bool
+    {
+        $test = $this->getInnerIterator()->current();
+
+        if ($test instanceof TestSuite) {
+            return true;
+        }
+
+        /** @var TestCase $test */
+        $testClass = get_class($test);
+
+        if (!isset($this->filter[$testClass])) {
+            return false;
+        }
+
+        $name = $test->getName();
+
+        return isset($this->filter[$testClass][$name]);
+    }
+
+    /**
+     * @throws Exception
+     */
+    private function setFilter(string $xmlFile): void
+    {
+        $xml = $this->parseXmlFromFile($xmlFile);
+
+        // Regular PHPUnit tests
+        foreach ($xml->testCaseClass as $class) {
+            $className = $class['name']->__toString();
```

Comment:
Can we make this typesafe? Maybe through an annotation like this:

```
/** @var array{name: ?} $class */
```

Reply:
Thanks for all the feedback!

I'll first wait regarding such typings as to in what direction the PR really goes.

----
I never used SimpleXml really before, this is all highly magic. `$class` is not an `array`, it's  `SimpleXMLElement` providing lots-o-magic stuff…

Reply:
I appreciate the time you invest in preparing this pull request. Please note, though, that I will not accept this if it introduces `ext/simplexml` usage.

Reply:
I've replaced SimpleXML with DomDocument and re-used the existing `\PHPUnit\Util\Xml\Loader::loadFile`